compiler/vmprofiler.nim

```diff
@@ -0,0 +1,42 @@
+
+import
+  options, vmdef, times, std/private/miscdollars, lineinfos, strutils, tables,
+  msgs
+  
+proc enter*(prof: var Profiler, c: PCtx, tos: PStackFrame) {.inline.} =
+  if optProfileVM in c.config.globalOptions:
+    prof.tEnter = cpuTime()
+    prof.tos = tos
+
+proc leave*(prof: var Profiler, c: PCtx) {.inline.} =
+  if optProfileVM in c.config.globalOptions:
+    let tLeave = cpuTime()
+    var tos = prof.tos
+    while tos != nil:
+      if tos.prc != nil:
+        let li = TLineInfo(fileIndex: tos.prc.info.fileIndex, line: tos.prc.info.line)
+        if li notin c.profiler.data:
+          c.profiler.data[li] = ProfileInfo()
+        c.profiler.data[li].time += tLeave - prof.tEnter
+        inc c.profiler.data[li].count
+      tos = tos.next
+
+proc dump*(p: var Profiler, c: PCtx) =
```

Comment:
general rule:
procs should almost never `echo`, instead it should return a string or (when performance dictates) take a `var string` argument; here I'd just use:
`proc dump*(p: var Profiler, c: PCtx): string =`

Reply:
True, but  'dump' only works if the `optProfileVM` is enabled. If I find the proper location to dump the end result I'll add the `if optProfileVM` there and change it into a string proc.

Reply:
here's how:
extract out `data` from `Profiler` and attach it to `ConfigRef`

```nim
# in compiler/options:
type ProfileInfo = ...
type ProfileData* = ref object
  data*: Table[TLineInfo, ProfileInfo]
  # other fields could be added, hence the wrapping ProfileData type

type ConfigRef = ...
  profileData*: ProfileData # add this field

proc dumpProfileData*(conf: ConfigRef): string # now you have all you need here
```

side benefit: it opens up possibility to reuse profiling for other parts of code (not just VM) since in the end the profiling is attached to a `TLineInfo`

(the VM code can have extra information thanks to `tos`, but other code could reuse some profiling logic)

then you can call `dumpProfileData` right before hintSuccessX rawMessage, and you'll have all the context there

Reply:
Right!

Reply:
done i guess